core/state_transition.go

```diff
@@ -351,6 +362,49 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 	activePrecompiles := vm.ActivePrecompiles(rules, st.evm.ChainConfig().Location)
 	st.state.PrepareAccessList(msg.From(), msg.To(), activePrecompiles, msg.AccessList(), st.evm.Config.Debug)
 
+	// If the transaction is coming from kQuaiSettingAddress, it can encode information in the data and
+	// perform three things
+	// 1) Set the KQuai exchange rate so that the protocol can use it going forward
+	// from the next block
+	// 2) Freeze the KQuai exchange rate computation, once this is done, the
+	// KQuai controller stops computing until its unforzen
+	// 3) UnFreeze the KQuai exchange rate computation, once this is done, the
+	// KQuai controller, starts computing the KQuai value
+	// This address cannot do anything else
+	if msg.From().Equal(kQuaiSettingAddress) && st.evm.Context.BlockNumber.Uint64() < params.BlocksPerYear {
+		if !st.msg.IsETX() && !contractCreation && bytes.Equal(st.data[:updateKQuaiLen], updateKQuai) {
```

Comment:
check st.data length as a sanity check otherwise this will panic

Reply:
true, will do